cf_remote/utils.py

```diff
@@ -226,15 +230,67 @@ def print_progress_dot(*args):
     sys.stdout.flush()  # STDOUT is line-buffered
 
 
+# atomic copy, see lock-free whack-a-mole algorithm
 def copy_file(input_path, output_path):
-    filename = os.path.basename(input_path)
-    output_dir = os.path.dirname(output_path)
 
-    tmp_filename = ".{}.tmp".format(filename)
-    tmp_output_path = os.path.join(output_dir, tmp_filename)
-
-    shutil.copyfile(input_path, tmp_output_path)
-    os.rename(tmp_output_path, output_path)
+    if not os.path.exists(input_path):
+        raise FileNotFoundError("File to copy '{}' doesn't exist".format(input_path))
+    if not os.path.isfile(input_path):
+        raise FileExistsError("'{}' exists and is not a file".format(input_path))
+
+    output_filename = os.path.basename(output_path)
+    if not output_filename:
+        output_filename = os.path.basename(input_path)
+
+    output_dirname = os.path.dirname(output_path)
+    my_uuid = uuid.uuid4()
+
+    tmp_fd, tmp_path = tempfile.mkstemp()
+    input_fd = os.open(input_path, os.O_RDONLY)
+
+    fcntl.flock(input_fd, fcntl.LOCK_EX)
+    shutil.copyfile(input_path, tmp_path)
+    fcntl.flock(input_fd, fcntl.LOCK_UN)
+
+    os.close(tmp_fd)
+    os.close(input_fd)
+
+    my_mole = os.path.join(
+        output_dirname, "{}-{}.mole.tmp".format(output_filename, my_uuid)
+    )
+    os.rename(tmp_path, my_mole)
+
+    glob_pattern = "{}-*.mole.tmp".format(output_filename)
+
+    def whack_a_mole(my_mole):
+        moles = glob.glob(os.path.join(output_dirname, glob_pattern))
+        for mole in moles:
+
+            if mole > my_mole:
```

Comment:
Maybe you could do something like this
```python
>>> fname_1 = "myfile.123.tmp.mole"
>>> fname_2 = "myfile.234.tmp.mole"
>>> file_to_wack = fname_1 if fname_1 < fname_2 else fname_2
>>> file_to_wack # Wack this one
'myfile.123.tmp.mole'
```
instead of having two almost identical blocks of code?